### ISSUE TITLE
3.0.2: `test_write_read_roundtrip` fails on some platforms

### ISSUE THREAD

**Issue by olebole** [astropy/astropy#7467]
On some Debian platforms, `test_write_read_roundtrip` fails. [On ARM 64 bit (little endian)](https://buildd.debian.org/status/fetch.php?pkg=astropy&arch=arm64&ver=3.0.2-1&stamp=1526381695&file=log):

```
__________________________ test_write_read_roundtrip ___________________________

    @pytest.mark.skipif('not HAS_YAML')
    def test_write_read_roundtrip():
        """[...]"""
        t = T_DTYPES
        for delimiter in DELIMITERS:
            out = StringIO()
            t.write(out, format='ascii.ecsv', delimiter=delimiter)
    
            t2s = [Table.read(out.getvalue(), format='ascii.ecsv'),
                    Table.read(out.getvalue(), format='ascii'),
                    ascii.read(out.getvalue()),
                    ascii.read(out.getvalue(), format='ecsv', guess=False),
                    ascii.read(out.getvalue(), format='ecsv')]
            for t2 in t2s:
                assert t.meta == t2.meta
                for name in t.colnames:
                    assert t[name].attrs_equal(t2[name])
>                   assert np.all(t[name] == t2[name])
E                   AssertionError: assert False
E                    +  where False = <function all at 0xffff8c021378>(<Column name=...3>\n0.0\n0.0\n0.0 == <Column name='...3>\n0.0\n0.0\n0.0
E                    +    where <function all at 0xffff8c021378> = np.all
E                       Full diff:
E                       <Column name='float128' dtype='float128' unit='m / s' description='descr_float128' length=3>
E                       0.0
E                       0.0
E                       0.0)

astropy/io/ascii/tests/test_ecsv.py:144: AssertionError
```

On Big Endian platforms, the failure is a bit different. [On s390x](https://buildd.debian.org/status/fetch.php?pkg=astropy&arch=s390x&ver=3.0.2-1&stamp=1526381141&file=log):
```
__________________________ test_write_read_roundtrip ___________________________

    @pytest.mark.skipif('not HAS_YAML')
    def test_write_read_roundtrip():
        """[...]"""
        t = T_DTYPES
        for delimiter in DELIMITERS:
        [...]
            for t2 in t2s:
                assert t.meta == t2.meta
                for name in t.colnames:
                    assert t[name].attrs_equal(t2[name])
>                   assert np.all(t[name] == t2[name])
E                   AssertionError: assert False
E                    +  where False = <function all at 0x3ffb744e2f0>(<Column name=...51571556046753 == <Column name='...           nan
E                    +    where <function all at 0x3ffb744e2f0> = np.all
E                       Full diff:
E                       <Column name='float128' dtype='float128' unit='m / s' description='descr_float128' length=3>
E                       -                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       0.0
E                       - 0.0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000016809489683107125218
E                       -  0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001681051571556046753
E                       +                    0.0
E                       + 5.12984914645603e-4937
E                       +                    nan)

astropy/io/ascii/tests/test_ecsv.py:144: AssertionError
```

This is Python 3.6.5, numpy 1.14.3, python-yaml 3.12, libyaml-0.1.7. Testing on x86 with an identical environment [succeeds](https://buildd.debian.org/status/fetch.php?pkg=astropy&arch=amd64&ver=3.0.2-1&stamp=1526380727&raw=0); also [ARM 32 bit](https://buildd.debian.org/status/fetch.php?pkg=astropy&arch=armhf&ver=3.0.2-1&stamp=1526389214&raw=0).

All build logs [here](https://buildd.debian.org/status/logs.php?pkg=astropy&ver=3.0.2-1).

With 3.0, this [did not happen](https://buildd.debian.org/status/logs.php?pkg=astropy&ver=3.0-3) (but was with Python 3.6.4 and numpy 1.13.3).

**Comment by pllim** [comment]
Is it possible for you to run it under Astropy 3.0 + Python 3.6.4 + numpy 1.14.3 combo? If it turns from success to failure, then this is caused by Numpy, not Astropy.

**Comment by olebole** [comment]
I tested astropy-3.0 in the current build environment (Python 3.6.5, numpy 1.14.3) and got the same failure on arm64. So, it is most probably an incompability with numpy. The question is however how to proceed now.

**Comment by pllim** [comment]
@mhvk , any advice?

**Comment by olebole** [comment]
It looks a bit mystic to me. I armed the code a bit:

```
       for t2 in t2s:
            assert t.meta == t2.meta
            for name in t.colnames:
                for x,y in zip(t[name], t2[name]):
                    assert x == y
                assert t[name].attrs_equal(t2[name])
                assert np.all(t[name] == t2[name]
```

and now I get (second data row, the column is `float128`):

```
                   for x,y in zip(t[name], t2[name]):
>                       assert x == y
E                       assert 0.0 == 0.0
astropy/io/ascii/tests/test_ecsv.py:148: AssertionError
```

So there are different zeroes in `float128`, and for some reason the first one is correct, but the second one differs.

**Comment by mhvk** [comment]
@olebole - most bizarre. There's also something very strange in the arm result. If I look at the initialization of `T_DTYPES`, every numerical column is initialized with `data = np.arange(3, dtype=dtype)`, so how come the column is all zeros?

Would it be possible to insert a `print(T_DTYPES)` and `print(out.getvalue())` statements  in the tests (just after `t.write(...)`? (pytest will capture that output and show it for failing tests.).

If it does turn out to be that `np.arange(3, dtype=np.float128)` just doesn't work, then it definitely is an upstream problem...

**Comment by olebole** [comment]
@mhvk Aahh, your careful observation probably solves the mystery. On arm64 I get:
```
>>> import numpy
>>> numpy.arange(3)
array([0, 1, 2])
>>> numpy.arange(3, dtype=numpy.float128)
array([0., 0., 0.], dtype=float128)
```
So, it is clear where to file a bug... Thanks for the hint!

**Comment by mhvk** [comment]
I'll re-open since this is an actual problem, but label it as requiring an upstream fix.

**Comment by pllim** [comment]
Please cross-link the relevant Numpy bug report here for future reference. Thank you!

**Comment by bsipocz** [comment]
It looks like this has been addressed upstream in  np 1.14.4


Feel free to reopen this or a follow-up if there are still issues.
